pkg/project/auth/cache.go

```diff
@@ -0,0 +1,482 @@
+			deleteSubjectsToNamespace(userSubjectRecordStore, reviewRecord.users, reviewRecord.namespace)
+			deleteSubjectsToNamespace(groupSubjectRecordStore, reviewRecord.groups, reviewRecord.namespace)
+			reviewRecordStore.Delete(reviewRecord)
+		}
+	}
+}
+
+// invalidateCache returns true if there was a change in the master namespace that holds global policy and policy bindings
+func (ac *AuthorizationCache) invalidateCache() bool {
+	invalidateCache := false
+
+	masterPolicies, err := ac.policyIndexer.Index("namespace", &authorizationapi.Policy{ObjectMeta: kapi.ObjectMeta{Namespace: ac.masterNamespace}})
+	if err != nil {
+		return true
+	}
+	for i := range masterPolicies {
+		policy := masterPolicies[i].(*authorizationapi.Policy)
+		if policy.ResourceVersion != ac.masterPolicyResourceVersion {
+			invalidateCache = true
+			ac.masterPolicyResourceVersion = policy.ResourceVersion
+		}
+	}
+
+	masterPolicyBindings, err := ac.policyBindingIndexer.Index("namespace", &authorizationapi.PolicyBinding{ObjectMeta: kapi.ObjectMeta{Namespace: ac.masterNamespace}})
+	if err != nil {
+		return true
+	}
+	for i := range masterPolicyBindings {
+		policyBinding := masterPolicyBindings[i].(*authorizationapi.PolicyBinding)
+		if policyBinding.ResourceVersion != ac.masterBindingResourceVersion {
+			invalidateCache = true
+			ac.masterBindingResourceVersion = policyBinding.ResourceVersion
+		}
+	}
+	return invalidateCache
+}
+
+// synchronize runs a a full synchronization over the cache data.  it must be run in a single-writer model, it's not thread-safe by design.
+func (ac *AuthorizationCache) synchronize() {
+	// TODO: upstream cache object should support a high-water mark, if there was no change in any of our caches, then we should be able to return quickly
+	skip := false
+	if skip {
+		return
+	}
+
+	// by default, we update our current caches and do an incremental change
+	userSubjectRecordStore := ac.userSubjectRecordStore
+	groupSubjectRecordStore := ac.groupSubjectRecordStore
+	reviewRecordStore := ac.reviewRecordStore
+
+	// if there was a global change that forced complete invalidation, we rebuild our cache and do a fast swap at end
+	invalidateCache := ac.invalidateCache()
+	if invalidateCache {
+		userSubjectRecordStore = cache.NewStore(subjectRecordKeyFunc)
+		groupSubjectRecordStore = cache.NewStore(subjectRecordKeyFunc)
+		reviewRecordStore = cache.NewStore(reviewRecordKeyFunc)
+	}
+
+	// iterate over caches and synchronize our three caches
+	namespaceSet := ac.synchronizeNamespaces(userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+	ac.synchronizePolicies(userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+	ac.synchronizePolicyBindings(userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+	purgeDeletedNamespaces(namespaceSet, userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+
+	// if we did a full rebuild, now we swap the fully rebuilt cache
+	if invalidateCache {
```

Comment:
As discussed, the cache is only invalidated at this scope if you change a master policy or policy binding.  Since that should be extremely rare, it should not invalidate in normal operation.  Given that, I am happy to go spin up 500 projects and let you know how long it takes for openshift to start and how much memory it consumes soon after this PR merges.